bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -822,10 +831,10 @@ public void run()
       };
     }
 
-    protected void release(boolean wait)
+    protected void release(boolean wait, boolean writer)
     {
       final int refCount = this.refCount.decrementAndGet();
-      if (refCount == 0 && storage != null) {
+      if (refCount == 0 && storage != null && canRelease(writer)) {
```

Comment:
DataList's memory manager should not be responsible for backpressure. It should be the responsibility of the BufferServer, this will allow us to change the policy of backpressure easily from Block granularity to Window/tuple granularity.

Instead of waiting till the end and blocking, adaptive sleep per tuple/window may be a good alternative, as it can quickly reduce the numbers of tuple generated per window.